pkg/util/time_cache_test.go

```diff
@@ -115,3 +117,57 @@ func TestCacheParallelOneCall(t *testing.T) {
 		t.Errorf("Expected %v, got %v", e, a)
 	}
 }
+
+func TestCacheParallelNoDeadlocksNoDoubleCalls(t *testing.T) {
+	keys := []string{}
+	fuzz.New().NilChance(0).NumElements(50, 50).Fuzz(&keys)
+	type callTrack struct {
+		sync.Mutex
+		accessTimes []time.Time
+	}
+	calls := map[string]*callTrack{}
+	for _, k := range keys {
+		calls[k] = &callTrack{}
+	}
+
+	ff := func(key string) T {
+		ct := calls[key]
+		ct.Lock()
+		ct.accessTimes = append(ct.accessTimes, time.Now())
+		ct.Unlock()
+		time.Sleep(time.Millisecond)
+		return key
+	}
+
+	clock := &FakeClock{time.Now()}
```

Comment:
(btw I wrote this test before writing the fix, and it was actually catching the bug even with a non-advancing clock.)